Add collapsible failed task logs to prevent React error 185

dheerajturaga opened the pull request:
This PR is an attempt to mitigate #52916. It does not "fix" the issue but prevents it from happening on the dag landing page.
- Recent failed task logs now start collapsed by default
- Added "Expand Logs"/"Hide Logs" toggle buttons with translations
- Implemented lazy loading - logs only fetch when expanded
- Prevents page crashes from large log content rendering
- Increased log container max height from 100px to 200px


https://github.com/user-attachments/assets/58436d00-b477-4493-b5d5-dc249c82fc48

<img width="1675" height="1067" alt="image" src="https://github.com/user-attachments/assets/f3c893ad-d2f6-4f31-9b06-da51ea751a0a" />

dheerajturaga committed:
Add collapsible failed task logs to prevent React error 185

- Recent failed task logs now start collapsed by default
- Added "Expand Logs"/"Hide Logs" toggle buttons with translations
- Implemented lazy loading - logs only fetch when expanded
- Prevents page crashes from large log content rendering
- Increased log container max height from 100px to 200px

bbovenzi commented on airflow-core/src/airflow/ui/public/i18n/locales/en/dag.json:
```diff
@@ -58,6 +58,8 @@
       "assetEvent_other": "Created Asset Events"
     },
     "failedLogs": {
+      "expandLogs": "Expand Logs",
```
```suggestion
      "showLogs": "Show Logs",
```

bbovenzi reviewed (commented):
Let's move over to show/hide terminology and avoid the confusion with expand/collapse log groups. Otherwise lgtm

dheerajturaga committed:
Apply suggestion from @bbovenzi

Co-authored-by: Brent Bovenzi <brent.bovenzi@gmail.com>

pierrejeambrun reviewed (commented):
If you try to do this with tasks that have big logs (10M) the collapsing will help, but after expanding a couple of them, the page still becomes unresponsive. (Without this, it's unresponsive/crash on load).

Yes, as mentionned by @guan404ming  ideally I think on the 'overview' we only want to show a log sample, maybe the last 100 lines of the logs to be able to debug most cases, if not enough they will need to go specifically check the TI logs on the TI page to load full content. For now we can just do this truncation in the frontend. (for instance custom hook wrapping useLogs and truncating response based on a param right after the backend response is received, i.e onSuccess, `useLogs(limit=100)`).

The long term solution would be to implement a backend feature to be able to paginate logs as well as a read order (start to end, end to start), so we could read the most recent 100 lines of the log file and return that directly, I don't think we support that yet cc: @jason810496

dheerajturaga committed:
Add optional  parameter to useLogs hook to truncate logs before rendering, preventing page unresponsiveness when expanding multiple large log previews. This addresses performance issues with tasks that have very large logs (10M+) while
  maintaining the lazy loading behavior.

  - Add limit parameter to useLogs hook Props interface
  - Implement log truncation logic using useMemo for performance
  - Update TaskLogPreview to use limit: 100 for overview display
  - Preserve backward compatibility - existing usage without limit unchanged

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -217,8 +220,23 @@ export const useLogs = (
   );
 
+  const truncatedData = useMemo(() => {
+    if (!data?.content || typeof limit !== "number" || limit <= 0) {
```
```suggestion
    if (!data?.content || limit === undefined || limit <= 0) {
```

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -217,8 +220,23 @@ export const useLogs = (
+    }
+
+    const streamingContent = parseStreamingLogContent(data);
+    const truncatedContent =
+      streamingContent.length > limit ? streamingContent.slice(-limit) : streamingContent;
+
```
Add a comment explaining why this is done in the front-end. (Backend does not support yet pagination / limits on logs reading endpoint)

dheerajturaga replied:
Done!

pierrejeambrun commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -217,8 +220,23 @@ export const useLogs = (
+
   const parsedData = parseLogs({
-    data: parseStreamingLogContent(data),
+    data: parseStreamingLogContent(typeof limit === "number" && limit > 0 ? truncatedData : data),
```
`truncatedData` already holds that logic.

```suggestion
    data: parseStreamingLogContent(truncatedData),
```

pierrejeambrun reviewed (commented):
A few suggestions before we can merge. Looking good otherwise.

dheerajturaga committed:
Pierre's Suggestions

pierrejeambrun reviewed (approved):
Thanks, ready to merge.
